compact_str/src/lib.rs

```diff
@@ -1170,12 +1173,98 @@ impl CompactString {
         let mut result = Self::with_capacity(v.len());
         let mut iter = v.iter();
         while let Some(s) = next_char(&mut iter, &mut buf) {
-            // SAFETY: next_char() only returns valid string
+            // SAFETY: next_char() only returns valid strings
             let s = unsafe { std::str::from_utf8_unchecked(s) };
             result.push_str(s);
         }
         result
     }
+
+    #[inline(always)]
+    fn from_utf16x(
+        v: &[u8],
+        from_x: impl Fn(u16) -> u16,
+        from_x_bytes: impl Fn([u8; 2]) -> u16,
+    ) -> Result<Self, Utf16Error> {
+        match unsafe { v.align_to::<u16>() } {
+            (&[], _, &[_]) | (&[_], _, &[]) => {
+                // Input had an odd number of bytes.
+                Err(Utf16Error(()))
+            }
+            (&[], v, _) => {
+                // Input is correcty aligned.
+                std::char::decode_utf16(v.iter().copied().map(from_x))
+                    .collect::<Result<CompactString, _>>()
+                    .map_err(|_| Utf16Error(()))
+            }
+            _ => {
+                // Input's alignment is off.
+                // SAFETY: we can always reinterpret a `[u8; 2*N]` slice as `[[u8; 2]; N]`
+                let v = unsafe { slice::from_raw_parts(v.as_ptr().cast(), v.len() / 2) };
+                std::char::decode_utf16(v.iter().copied().map(from_x_bytes))
+                    .collect::<Result<CompactString, _>>()
+                    .map_err(|_| Utf16Error(()))
+            }
+        }
+    }
+
+    /// Decode a slice of bytes as UTF-16 encoded string, in little endian.
+    ///
+    /// # Errors
+    ///
+    /// If the slice has an odd number of bytes, or if it did not contain valid UTF-16 characters,
+    /// a [`Utf16Error`] is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use compact_str::CompactString;
+    /// const DANCING_MEN: &[u8] = b"\x3d\xd8\x6f\xdc\x0d\x20\x42\x26\x0f\xfe";
+    /// let dancing_men = CompactString::from_utf16le(DANCING_MEN).unwrap();
+    /// assert_eq!(dancing_men, "👯‍♂️");
+    /// ```
+    pub fn from_utf16le(v: &[u8]) -> Result<Self, Utf16Error> {
```

Comment:
nit: can this be generic over some type that implements `AsRef<[u8]>`, e.g. `from_utf16le<B: AsRef<[u8]>>(buf: B)` to be consistent with the other `from_*(...)` APIs

Reply:
Made generic.